File: modules/adapted/raptorbt/test_fill_timing.py
"""Execution-timing causality, exercised through the wheel.

In plain words: a decision made from a bar's data may trade no earlier than
that bar's own close. ``fill_timing="next_bar_open"`` executes it at the
next bar's open. Through 0.10, ``upon_bar_close=False`` filled the decision
at the SAME bar's open — a price from before the decision's information
existed — and that behavior is now only reachable by explicitly asking for
``"same_bar_open_lookahead"``.
"""

import numpy as np
import pytest

import raptorbt
from raptorbt import BacktestConfig


def _rally_fixture(n=12):
    """One bar rallies 100 -> 200 intrabar; every other price is 150.

    A causal trader learns "bar 5 closed at 200" only when it closes; the
    cheapest price available from then on is 150 and the run ends at 150,
    so no causal strategy acting on this signal can make money.
    """
    ts = np.arange(n, dtype=np.int64) * 1_000_000_000
    o = np.full(n, 150.0)
    h = np.full(n, 150.0)
    l = np.full(n, 150.0)
    c = np.full(n, 150.0)
    o[5], l[5], h[5], c[5] = 100.0, 100.0, 200.0, 200.0
    v = np.full(n, 1000.0)
    entries = np.zeros(n, dtype=bool)
    exits = np.zeros(n, dtype=bool)
    entries[5] = True  # emitted because close[5] == 200
    exits[10] = True
    return ts, o, h, l, c, v, entries, exits


def _run(config):
    ts, o, h, l, c, v, entries, exits = _rally_fixture()
    return raptorbt.run_single_backtest(
        ts, o, h, l, c, v, entries, exits, direction=1, config=config
    )


def _no_cost_config(**kwargs):
    return BacktestConfig(fees=0.0, slippage=0.0, **kwargs)


class TestNextBarOpen:
    def test_fills_the_bar_after_the_signal(self):
        result = _run(_no_cost_config(fill_timing="next_bar_open"))
        trades = result.trades()
        assert len(trades) == 1
        # Decision at bar 5 (close 200) fills at bar 6's open (150).
        assert trades[0].entry_idx == 6
        assert trades[0].entry_price == 150.0
        # The causally-impossible ~+50% is gone.
        assert abs(result.metrics.total_return_pct) < 1e-9

    def test_deprecated_bool_false_means_next_bar_open(self):
        result = _run(_no_cost_config(upon_bar_close=False))
        assert result.trades()[0].entry_price == 150.0
        assert result.trades()[0].entry_idx == 6

    def test_explicit_fill_timing_wins_over_the_bool(self):
        cfg = _no_cost_config(upon_bar_close=True, fill_timing="next_bar_open")
        result = _run(cfg)
        assert result.trades()[0].entry_idx == 6


class TestLegacyAndDefaults:
    def test_lookahead_mode_reproduces_pre_0_11_results_by_name_only(self):
        result = _run(_no_cost_config(fill_timing="same_bar_open_lookahead"))
        trade = result.trades()[0]
        assert trade.entry_idx == 5
        assert trade.entry_price == 100.0  # the pre-0.11 look-ahead fill
        assert result.metrics.total_return_pct > 45.0

    def test_default_same_bar_close_is_unchanged(self):
        result = _run(_no_cost_config())
        trade = result.trades()[0]
        assert trade.entry_idx == 5
        assert trade.entry_price == 200.0  # the decision bar's close


class TestConfigSurface:
    def test_invalid_fill_timing_is_refused_loudly(self):
        with pytest.raises(ValueError, match="fill_timing"):
            BacktestConfig(fill_timing="next_open")

    def test_fill_timing_reads_back(self):
        assert BacktestConfig().fill_timing is None
        assert (
            BacktestConfig(fill_timing="next_bar_open").fill_timing
            == "next_bar_open"
        )


class TestPremiumOpenSeries:
    """Optional open premiums for the premium-only runners.

    Under fill_timing="next_bar_open" a signal fill prices at the fill
    bar's OPEN premium when the caller supplies the series; without it the
    bar's settled value remains the fill. Nothing is synthesized.
    """

    def _spot(self, n=10):
        ts = np.arange(n, dtype=np.int64) * 1_000_000_000
        o = np.full(n, 100.0)
        h = np.full(n, 101.0)
        l = np.full(n, 99.0)
        c = np.full(n, 100.0)
        v = np.full(n, 1000.0)
        return ts, o, h, l, c, v

    def test_options_fill_at_the_open_premium(self):
        ts, o, h, l, c, v = self._spot()
        premiums = 10.0 + np.arange(10, dtype=np.float64)
        opens = 100.0 + np.arange(10, dtype=np.float64)
        entries = np.zeros(10, dtype=bool)
        exits = np.zeros(10, dtype=bool)
        entries[3] = True
        exits[6] = True

        result = raptorbt.run_options_backtest(
            ts, o, h, l, c, v, premiums, entries, exits,
            direction=1,
            config=_no_cost_config(fill_timing="next_bar_open"),
            option_open_prices=opens,
        )
        trade = result.trades()[0]
        assert trade.entry_idx == 4
        assert trade.entry_price == 104.0  # bar 4's OPEN premium
        assert trade.exit_price == 107.0

    def test_options_open_series_length_is_validated(self):
        ts, o, h, l, c, v = self._spot()
        premiums = np.full(10, 10.0)
        with pytest.raises(ValueError, match="option_open_prices"):
            raptorbt.run_options_backtest(
                ts, o, h, l, c, v, premiums,
                np.zeros(10, dtype=bool), np.zeros(10, dtype=bool),
                option_open_prices=np.full(9, 10.0),
            )

    def test_spread_fills_legs_at_their_open_premiums(self):
        ts, _, _, _, c, _ = self._spot()
        legs = [10.0 + np.arange(10, dtype=np.float64),
                20.0 + np.arange(10, dtype=np.float64)]
        opens = [100.0 + np.arange(10, dtype=np.float64),
                 200.0 + np.arange(10, dtype=np.float64)]
        entries = np.zeros(10, dtype=bool)
        exits = np.zeros(10, dtype=bool)
        entries[3] = True
        exits[6] = True

        result = raptorbt.run_spread_backtest(
            ts, c, legs,
            [("CE", 100.0, 1, 1), ("PE", 100.0, 1, 1)],
            entries, exits,
            config=_no_cost_config(fill_timing="next_bar_open"),
            spread_type="straddle",
            legs_open_premiums=opens,
        )
        trade = result.trades()[0]
        assert trade.entry_idx == 4
        assert trade.exit_idx == 7
        # Long both legs, open-premium fills: +3 per leg. Settled-value
        # fills would report +6 per leg.
        assert abs(trade.pnl - 6.0) < 1e-9

    def test_spread_open_series_shape_is_validated(self):
        ts, _, _, _, c, _ = self._spot()
        legs = [np.full(10, 10.0), np.full(10, 20.0)]
        with pytest.raises(ValueError, match="legs_open_premiums"):
            raptorbt.run_spread_backtest(
                ts, c, legs,
                [("CE", 100.0, 1, 1), ("PE", 100.0, 1, 1)],
                np.zeros(10, dtype=bool), np.zeros(10, dtype=bool),
                legs_open_premiums=[np.full(10, 10.0)],  # one leg missing
            )

File: modules/adapted/raptorbt/test_fill_timing.rs
//! Guard tests for execution-timing causality.
//!
//! A decision made from bar i's data may trade no earlier than bar i's own
//! close. Under `FillTiming::NextBarOpen` it trades at bar i+1's open.
//! Through 0.10, `upon_bar_close=false` filled a bar-i signal at bar i's
//! OWN open — a price that traded before the information the signal used
//! existed. These tests pin the corrected semantics; if the shortcut ever
//! comes back, they fail.

use raptorbt::core::types::{
    BacktestConfig, CompiledSignals, Direction, ExitReason, FillTiming, OhlcvData, StopConfig,
    TargetConfig,
};
use raptorbt::portfolio::engine::PortfolioEngine;

fn config(fill_timing: FillTiming) -> BacktestConfig {
    BacktestConfig {
        initial_capital: 100_000.0,
        fees: 0.0,
        slippage: 0.0,
        stop: StopConfig::None,
        target: TargetConfig::None,
        fill_timing: Some(fill_timing),
        ..Default::default()
    }
}

fn flat_ohlcv(n: usize, price: f64) -> OhlcvData {
    OhlcvData {
        timestamps: (0..n as i64).map(|i| i * 1_000_000_000).collect(),
        open: vec![price; n],
        high: vec![price; n],
        low: vec![price; n],
        close: vec![price; n],
        volume: vec![1000.0; n],
    }
}

fn signals(n: usize, entry_at: usize, exit_at: Option<usize>) -> CompiledSignals {
    let mut entries = vec![false; n];
    let mut exits = vec![false; n];
    entries[entry_at] = true;
    if let Some(i) = exit_at {
        exits[i] = true;
    }
    CompiledSignals {
        symbol: "GUARD".to_string(),
        entries,
        exits,
        position_sizes: None,
        direction: Direction::Long,
        weight: 1.0,
    }
}

/// One bar rallies 100 -> 200 intrabar; every later price is 150 forever.
///
/// A causal trader learns "bar 5 closed at 200" only when it closes; the
/// cheapest price available from that moment on is 150, and the run ends at
/// 150 — no causal strategy acting on this signal can make money.
fn rally_fixture() -> OhlcvData {
    let mut ohlcv = flat_ohlcv(12, 150.0);
    ohlcv.open[5] = 100.0;
    ohlcv.low[5] = 100.0;
    ohlcv.high[5] = 200.0;
    ohlcv.close[5] = 200.0;
    // Bars before the signal sit at 150 too.
    for i in 0..5 {
        ohlcv.open[i] = 150.0;
        ohlcv.high[i] = 150.0;
        ohlcv.low[i] = 150.0;
        ohlcv.close[i] = 150.0;
    }
    ohlcv
}

#[test]
fn next_bar_open_fills_the_bar_after_the_signal() {
    let engine = PortfolioEngine::new(config(FillTiming::NextBarOpen));
    let result = engine.run_single(&rally_fixture(), &signals(12, 5, Some(10)));

    assert_eq!(result.trades.len(), 1);
    let trade = &result.trades[0];

    // The signal is a function of bar 5's close (200); the earliest
    // tradeable price after that is bar 6's open.
    assert_eq!(trade.entry_idx, 6, "entry fills on the bar after the signal");
    assert_eq!(trade.entry_price, 150.0, "entry fills at the next bar's open");
    // The exit signal at bar 10 fills at bar 11's open.
    assert_eq!(trade.exit_idx, 11);
    assert_eq!(trade.exit_price, 150.0);

    // The causally-impossible ~+50% is gone: buy 150, sell 150.
    assert!(
        result.metrics.total_return_pct.abs() < 1e-9,
        "no causal trader earns anything here; reported {}%",
        result.metrics.total_return_pct
    );
}

#[test]
fn next_bar_open_books_the_crash_the_exit_reacted_to() {
    // Crash bar: opens 150, closes 50; price stays at 50 afterwards. The
    // exit signal exists *because* of the 50 close, so a causal trader
    // cannot get out above 50.
    let mut ohlcv = flat_ohlcv(12, 150.0);
    ohlcv.low[8] = 50.0;
    ohlcv.close[8] = 50.0;
    for i in 9..12 {
        ohlcv.open[i] = 50.0;
        ohlcv.high[i] = 50.0;
        ohlcv.low[i] = 50.0;
        ohlcv.close[i] = 50.0;
    }

    let engine = PortfolioEngine::new(config(FillTiming::NextBarOpen));
    let result = engine.run_single(&ohlcv, &signals(12, 2, Some(8)));

    let trade = &result.trades[0];
    assert_eq!(trade.entry_idx, 3, "entered at the bar after the entry signal");
    assert_eq!(trade.entry_price, 150.0);
    assert_eq!(trade.exit_idx, 9, "exit fills the bar after the crash was seen");
    assert_eq!(trade.exit_price, 50.0, "the crash is not dodged");
    assert!(trade.pnl < 0.0, "the loss is real and must be booked");
}

/// The explicitly-named legacy mode reproduces the pre-0.11 numbers, and
/// nothing else does.
#[test]
fn lookahead_mode_reproduces_pre_0_11_results_by_name_only() {
    let engine = PortfolioEngine::new(config(FillTiming::SameBarOpenLookahead));
    let result = engine.run_single(&rally_fixture(), &signals(12, 5, Some(10)));

    let trade = &result.trades[0];
    // Pre-0.11 `upon_bar_close=false`: entry at the signal bar's own open.
    assert_eq!(trade.entry_idx, 5);
    assert_eq!(trade.entry_price, 100.0);
    assert_eq!(trade.exit_price, 150.0);
    assert!(
        result.metrics.total_return_pct > 45.0,
        "the legacy mode must replay the old (unearnable) result exactly"
    );
}

/// The deprecated bool maps onto the corrected semantics: `false` is
/// next-bar-open, never the old same-bar-open look-ahead.
#[test]
fn deprecated_bool_resolves_to_next_bar_open() {
    let cfg = BacktestConfig {
        initial_capital: 100_000.0,
        fees: 0.0,
        slippage: 0.0,
        stop: StopConfig::None,
        target: TargetConfig::None,
        upon_bar_close: false,
        fill_timing: None,
        ..Default::default()
    };
    assert_eq!(cfg.resolved_fill_timing(), FillTiming::NextBarOpen);

    let result = PortfolioEngine::new(cfg).run_single(&rally_fixture(), &signals(12, 5, Some(10)));
    assert_eq!(result.trades[0].entry_price, 150.0, "bool=false must not reach bar 5's open");
}

/// A signal on the final bar has no next bar to fill on: it does not trade.
#[test]
fn final_bar_signal_never_fills() {
    let engine = PortfolioEngine::new(config(FillTiming::NextBarOpen));
    let result = engine.run_single(&flat_ohlcv(12, 150.0), &signals(12, 11, None));
    assert!(result.trades.is_empty(), "a last-bar decision has nothing to trade against");
}

/// A position opened at bar i+1's open lives through bar i+1: its stop can
/// fire within the same bar it was filled on.
#[test]
fn deferred_entry_can_stop_out_within_its_fill_bar() {
    let mut cfg = config(FillTiming::NextBarOpen);
    cfg.stop = StopConfig::Fixed { percent: 0.02 };

    // Signal at bar 4; bar 5 opens at 150 and collapses through the stop.
    let mut ohlcv = flat_ohlcv(12, 150.0);
    ohlcv.low[5] = 100.0;
    ohlcv.close[5] = 110.0;
    for i in 6..12 {
        ohlcv.open[i] = 110.0;
        ohlcv.high[i] = 110.0;
        ohlcv.low[i] = 110.0;
        ohlcv.close[i] = 110.0;
    }

    let result = PortfolioEngine::new(cfg).run_single(&ohlcv, &signals(12, 4, None));
    let trade = &result.trades[0];
    assert_eq!(trade.entry_idx, 5);
    assert_eq!(trade.entry_price, 150.0);
    assert_eq!(trade.exit_reason, ExitReason::StopLoss);
    assert_eq!(trade.exit_idx, 5, "the fill bar's own range applies to the new position");
    assert_eq!(trade.exit_price, 147.0, "2% below the 150.0 entry");
}

/// Same-bar-close mode is untouched by the fix: a signal at bar i still
/// fills at bar i's close, the price the decision coincides with.
#[test]
fn same_bar_close_semantics_are_unchanged() {
    let engine = PortfolioEngine::new(config(FillTiming::SameBarClose));
    let result = engine.run_single(&rally_fixture(), &signals(12, 5, Some(10)));

    let trade = &result.trades[0];
    assert_eq!(trade.entry_idx, 5);
    assert_eq!(trade.entry_price, 200.0, "close-mode fills at the decision bar's close");
}

// ---- Multi-leg runners -------------------------------------------------
//
// Basket and pairs legs carry full OHLC, so NextBarOpen fills at each
// leg's own open on the bar after the decision. Their default (and
// historical) behavior is same-bar-close, which stays byte-identical.

fn stepped_ohlcv(n: usize) -> OhlcvData {
    // Distinct open vs close on every bar so a wrong price source is
    // always visible: bar i opens at 100+i and closes at 200+i.
    OhlcvData {
        timestamps: (0..n as i64).map(|i| i * 1_000_000_000).collect(),
        open: (0..n).map(|i| 100.0 + i as f64).collect(),
        high: (0..n).map(|i| 300.0 + i as f64).collect(),
        low: (0..n).map(|i| 90.0 + i as f64).collect(),
        close: (0..n).map(|i| 200.0 + i as f64).collect(),
        volume: vec![1000.0; n],
    }
}

#[test]
fn basket_next_bar_open_fills_every_leg_at_its_next_open() {
    use raptorbt::strategies::basket::{BasketBacktest, BasketConfig};

    let cfg = BasketConfig { base: config(FillTiming::NextBarOpen), ..BasketConfig::default() };
    let instruments = vec![
        (stepped_ohlcv(10), signals(10, 3, Some(6))),
        (stepped_ohlcv(10), signals(10, 3, Some(6))),
    ];
    let result = BasketBacktest::new(cfg).run(&instruments);

    assert_eq!(result.trades.len(), 2, "one trade per leg");
    for trade in &result.trades {
        assert_eq!(trade.entry_idx, 4, "decision at 3 fills on 4");
        assert_eq!(trade.entry_price, 104.0, "at bar 4's open");
        assert_eq!(trade.exit_idx, 7, "decision at 6 fills on 7");
        assert_eq!(trade.exit_price, 107.0, "at bar 7's open");
    }
}

#[test]
fn basket_default_close_fills_are_unchanged() {
    use raptorbt::strategies::basket::{BasketBacktest, BasketConfig};

    let cfg = BasketConfig { base: config(FillTiming::SameBarClose), ..BasketConfig::default() };
    let instruments = vec![(stepped_ohlcv(10), signals(10, 3, Some(6)))];
    let result = BasketBacktest::new(cfg).run(&instruments);

    let trade = &result.trades[0];
    assert_eq!((trade.entry_idx, trade.entry_price), (3, 203.0));
    assert_eq!((trade.exit_idx, trade.exit_price), (6, 206.0));
}

#[test]
fn basket_last_bar_signal_never_fills() {
    use raptorbt::strategies::basket::{BasketBacktest, BasketConfig};

    let cfg = BasketConfig { base: config(FillTiming::NextBarOpen), ..BasketConfig::default() };
    let instruments = vec![(stepped_ohlcv(10), signals(10, 9, None))];
    let result = BasketBacktest::new(cfg).run(&instruments);
    assert!(result.trades.is_empty(), "a last-bar decision has nothing to trade against");
}

#[test]
fn pairs_next_bar_open_fills_both_legs_at_their_next_open() {
    use raptorbt::strategies::pairs::{PairsBacktest, PairsConfig};

    let cfg = PairsConfig { base: config(FillTiming::NextBarOpen), ..PairsConfig::default() };
    let leg1 = stepped_ohlcv(10);
    let mut leg2 = stepped_ohlcv(10);
    // Different open levels per leg so each leg's own series is provably used.
    for v in leg2.open.iter_mut() {
        *v += 50.0;
    }
    let result = PairsBacktest::new(cfg).run(&leg1, &leg2, &signals(10, 3, Some(6)));

    assert_eq!(result.trades.len(), 2);
    let l1 = &result.trades[0];
    let l2 = &result.trades[1];
    assert_eq!((l1.entry_idx, l1.entry_price), (4, 104.0), "leg1 at its bar-4 open");
    assert_eq!((l2.entry_idx, l2.entry_price), (4, 154.0), "leg2 at its bar-4 open");
    assert_eq!(l1.exit_price, 107.0);
    assert_eq!(l2.exit_price, 157.0);
}

#[test]
fn pairs_default_close_fills_are_unchanged() {
    use raptorbt::strategies::pairs::{PairsBacktest, PairsConfig};

    let cfg = PairsConfig { base: config(FillTiming::SameBarClose), ..PairsConfig::default() };
    let leg = stepped_ohlcv(10);
    let result = PairsBacktest::new(cfg).run(&leg, &leg.clone(), &signals(10, 3, Some(6)));
    assert_eq!(result.trades[0].entry_price, 203.0);
    assert_eq!(result.trades[0].exit_price, 206.0);
}

// Options and spread legs are premium-only series — no open exists — so
// NextBarOpen fills at the bar AFTER the decision, at that bar's premium
// (the honest causal fill). Strike selection stays on the decision bar.

#[test]
fn options_next_bar_open_fills_at_the_next_bars_premium() {
    use raptorbt::strategies::options::{OptionsBacktest, OptionsConfig};

    let spot = stepped_ohlcv(10);
    // Premium series: distinct per bar (10 + i).
    let premiums: Vec<f64> = (0..10).map(|i| 10.0 + i as f64).collect();

    let cfg = OptionsConfig { base: config(FillTiming::NextBarOpen), ..OptionsConfig::default() };
    let result = OptionsBacktest::new(cfg).run(&spot, &premiums, &signals(10, 3, Some(6)));

    let trade = &result.trades[0];
    assert_eq!(trade.entry_idx, 4, "decision at 3 fills on 4");
    assert_eq!(trade.entry_price, 14.0, "at bar 4's premium");
    assert_eq!(trade.exit_idx, 7);
    assert_eq!(trade.exit_price, 17.0, "at bar 7's premium");

    // Default mode unchanged: same-bar premium.
    let cfg = OptionsConfig { base: config(FillTiming::SameBarClose), ..OptionsConfig::default() };
    let result = OptionsBacktest::new(cfg).run(&spot, &premiums, &signals(10, 3, Some(6)));
    assert_eq!(result.trades[0].entry_price, 13.0);
    assert_eq!(result.trades[0].exit_price, 16.0);
}

#[test]
fn spreads_next_bar_open_prices_legs_off_the_next_bar() {
    use raptorbt::strategies::{
        LegConfig, SpreadBacktest, SpreadConfig, SpreadOptionType, SpreadType,
    };

    let n = 10;
    let timestamps: Vec<i64> = (0..n as i64).map(|i| i * 1_000_000_000).collect();
    let underlying: Vec<f64> = vec![100.0; n];
    // Two legs with distinct per-bar premiums.
    let legs_premiums = vec![
        (0..n).map(|i| 10.0 + i as f64).collect::<Vec<f64>>(),
        (0..n).map(|i| 20.0 + i as f64).collect::<Vec<f64>>(),
    ];
    let mut entries = vec![false; n];
    let mut exits = vec![false; n];
    entries[3] = true;
    exits[6] = true;

    let make = |timing: FillTiming| SpreadConfig {
        base: config(timing),
        spread_type: SpreadType::Straddle,
        leg_configs: vec![
            LegConfig::new(SpreadOptionType::Call, 100.0, 1, 1),
            LegConfig::new(SpreadOptionType::Put, 100.0, -1, 1),
        ],
        max_loss: None,
        target_profit: None,
        leg_expiry_timestamps: None,
    };

    let result = SpreadBacktest::new(make(FillTiming::NextBarOpen)).run(
        &timestamps,
        &underlying,
        &legs_premiums,
        &entries,
        &exits,
    );
    assert_eq!(result.trades.len(), 1);
    let trade = &result.trades[0];
    assert_eq!(trade.entry_idx, 4, "decision at 3 opens on 4, at bar 4's premiums");
    assert_eq!(trade.exit_idx, 7, "decision at 6 closes on 7");

    // Default mode unchanged: same-bar entry/exit indices.
    let result = SpreadBacktest::new(make(FillTiming::SameBarClose)).run(
        &timestamps,
        &underlying,
        &legs_premiums,
        &entries,
        &exits,
    );
    assert_eq!(result.trades[0].entry_idx, 3);
    assert_eq!(result.trades[0].exit_idx, 6);
}

#[test]
fn options_next_bar_open_uses_the_open_premium_when_supplied() {
    use raptorbt::strategies::options::{OptionsBacktest, OptionsConfig};

    let spot = stepped_ohlcv(10);
    // Settled premiums 10+i; opening premiums 100+i — distinct on every bar
    // so the wrong series is always visible.
    let premiums: Vec<f64> = (0..10).map(|i| 10.0 + i as f64).collect();
    let opens: Vec<f64> = (0..10).map(|i| 100.0 + i as f64).collect();

    let cfg = OptionsConfig { base: config(FillTiming::NextBarOpen), ..OptionsConfig::default() };
    let result = OptionsBacktest::new(cfg).run_with_opens(
        &spot,
        &premiums,
        Some(&opens),
        &signals(10, 3, Some(6)),
    );

    let trade = &result.trades[0];
    assert_eq!(trade.entry_idx, 4);
    assert_eq!(trade.entry_price, 104.0, "the fill bar's OPEN premium, not its settled value");
    assert_eq!(trade.exit_idx, 7);
    assert_eq!(trade.exit_price, 107.0);

    // Outside NextBarOpen the opens are ignored: same-bar fills coincide
    // with the decision bar's settled value by design.
    let cfg = OptionsConfig { base: config(FillTiming::SameBarClose), ..OptionsConfig::default() };
    let result = OptionsBacktest::new(cfg).run_with_opens(
        &spot,
        &premiums,
        Some(&opens),
        &signals(10, 3, Some(6)),
    );
    assert_eq!(result.trades[0].entry_price, 13.0);
}

#[test]
fn spreads_next_bar_open_prices_signal_fills_at_leg_open_premiums() {
    use raptorbt::strategies::{
        LegConfig, SpreadBacktest, SpreadConfig, SpreadOptionType, SpreadType,
    };

    let n = 10;
    let timestamps: Vec<i64> = (0..n as i64).map(|i| i * 1_000_000_000).collect();
    let underlying = vec![100.0; n];
    // Settled premiums vs opening premiums, distinct per bar and per leg.
    let legs_premiums = vec![
        (0..n).map(|i| 10.0 + i as f64).collect::<Vec<f64>>(),
        (0..n).map(|i| 20.0 + i as f64).collect::<Vec<f64>>(),
    ];
    let legs_opens = vec![
        (0..n).map(|i| 100.0 + i as f64).collect::<Vec<f64>>(),
        (0..n).map(|i| 200.0 + i as f64).collect::<Vec<f64>>(),
    ];
    let mut entries = vec![false; n];
    let mut exits = vec![false; n];
    entries[3] = true;
    exits[6] = true;

    let cfg = SpreadConfig {
        base: config(FillTiming::NextBarOpen),
        spread_type: SpreadType::Straddle,
        leg_configs: vec![
            LegConfig::new(SpreadOptionType::Call, 100.0, 1, 1),
            LegConfig::new(SpreadOptionType::Put, 100.0, 1, 1),
        ],
        max_loss: None,
        target_profit: None,
        leg_expiry_timestamps: None,
    };
    let result = SpreadBacktest::new(cfg).run_with_opens(
        &timestamps,
        &underlying,
        &legs_premiums,
        Some(&legs_opens),
        &entries,
        &exits,
    );

    assert_eq!(result.trades.len(), 1);
    let trade = &result.trades[0];
    assert_eq!(trade.entry_idx, 4, "decision at 3 opens on 4");
    assert_eq!(trade.exit_idx, 7, "decision at 6 closes on 7");
    // Long both legs: entry pays open[4] per leg (104 + 204), signal exit
    // receives open[7] per leg (107 + 207) — a P&L of +3 per leg, from the
    // OPEN premiums. Marked-value fills (11+21 -> 17+27) would report +6
    // per leg instead, so the pnl pins which series filled.
    assert!(
        (trade.pnl - 6.0).abs() < 1e-9,
        "P&L must come from the open premiums (+3 per leg), got {}",
        trade.pnl
    );
}
